functions/src/utils/agora.js

```diff
@@ -0,0 +1,41 @@
+const functions = require("firebase-functions");
+
+const { RtcTokenBuilder, RtcRole } = require("agora-access-token");
+
+const AGORA_CONFIG = functions.config().agora;
+
+const appId = AGORA_CONFIG.app_id;
+const appCertificate = AGORA_CONFIG.app_certificate;
+
+// @debt should we move this into AGORA_CONFIG, or maybe venue config in firestore or similar?
+const expirationTimeInSeconds = 3600;
+
+const getExpirationTime = () => {
+  const currentTimestamp = Math.floor(Date.now() / 1000);
+
+  return currentTimestamp + expirationTimeInSeconds;
+};
+
+const generateAgoraTokenForAccount = ({ channelName, account, role }) => {
+  // @debt we should enforce a stricter security requirement on channelName. Maybe use UUIDs?
+  if (typeof channelName !== "string" && channelName.length <= 0)
+    throw new Error("channelName must be a string");
```

Comment:
> we should enforce a stricter security requirement on channelName. Maybe use UUIDs?

[OOS] This is probably OOS for this PR, as it would require deeper integration with the rest of the frontend code, which isn't available on `staging` currently

Reply:
> we should enforce a stricter security requirement on channelName. Maybe use UUIDs?

(another option) I guess channel name can be also generated using some hash function based on venueId.

Reply:
That's roughly close to what we do currently with Twilio, but i'm not super happy about it. If we do use a hash function, it would need to use secret data within the hash, which would need to be unguessable (since our code is open source, and venueId is a known value, 'security through obscurity' wouldn't be enough here.

Since we would need to store a secret value in firestore/etc anyways, we may as well just make that the whole value we use IMO.

Reply:
(For continuity, see also this thread: https://github.com/sparkletown/sparkle/pull/1673#discussion_r659327633 on the same topic)

Reply:
aha, I see. Makes sense. yep, forgot a little that we're open source. 😄

Reply:
Hehe. Even when not open source, security through obscurity isn't a great idea. Me and my friends have made a lot of bug bounty (ethical hacking) money from projects/companies that have relied on less than secure implementations :p

Reply:
hahaha woooooow! 😲 Super cooool!

Where did you get all this security knowledge? Any resource that comes to your mind first(for newbies)? Or it's just accumulated by your experience? I remember you were at a security conference some time ago. Sounds like a really captivating topic to learn deeper.

Reply:
Mostly a lifetime of interest/being 'jacked in' to the internet + an insatiable curiosity for wanting to know how things work, and then how to make them work how I want them to rather than how they're meant to. Then i've bounced back and forth between dev and security as I burn out/get bored of one or the other 😛 

Hit me up on Slack/etc about it sometime when deadlines aren't as pressing and let's chat a bit more. I can likely dig up some good starting points/etc for you, depending on what level of interest/how deep you want to go/etc. :)

Reply:
THAT
WOULD
BE
AWESOMEEE!!! 

Appreciate your will to help. Thanks!!! ❤️